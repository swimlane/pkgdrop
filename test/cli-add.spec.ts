import { createSandbox } from './cli.util';

const TIMEOUT = 100000;

describe('add', () => {
  let output: string;
  let sandbox: any;

  beforeAll(async () => {
    sandbox = await createSandbox();
    await sandbox.exec(`init -y`);
    output = await sandbox.exec(`add lit-element@2.1.0 --clean`);
  }, TIMEOUT);

  afterAll(async () => {
    await sandbox.clean();
  });

  afterEach(() => {
    process.exitCode = null;
  });

  test('displays console messages', async () => {
    expect(output).toMatchSnapshot();
    expect(process.exitCode).toBeFalsy();
  });

  test('files exist', async () => {
    expect(await sandbox.exists('-/lit-element@2.1.0')).toBe('dir');
    expect(await sandbox.exists('-/lit-html@1.4.1')).toBe('dir'); // Note: changes to latest v1
    expect(await sandbox.exists('-/importmap.json')).toBe('file');
  });

  test('importmap', async () => {
    expect(await sandbox.read('-/importmap.json')).toMatchSnapshot();
  });

  test(`can't add again`, async () => {
    const out = await sandbox.exec(`add lit-element@2.1.0`);
    expect(out).toContain('Package lit-element@2.1.0 already exists, skipping');
    expect(process.exitCode).toBeFalsy();
  });

  test('can force', async () => {
    const out = await sandbox.exec(`add lit-element@2.1.0 --force`);
    expect(out).toMatchSnapshot();
    expect(process.exitCode).toBeFalsy();
  });

  test('fail on missing package', async () => {
    const out = await sandbox.exec(`add @swimlane/pkgdrop@0.0.0`);
    expect(out).toMatchSnapshot();
    expect(process.exitCode).toBeTruthy(); // exit code 1
  });

  test('can force skip on missing packages', async () => {
    const out = await sandbox.exec(`add @swimlane/pkgdrop@0.0.0 --force`);
    expect(out).toMatchSnapshot();
    expect(process.exitCode).toBeFalsy();
  });

  test('can clean', async () => {
    const out = await sandbox.exec(`add lit-html@1.1.0 --clean`);
    expect(out).toMatchSnapshot();
    expect(await sandbox.exists('-/lit-element@2.1.0')).toBe(false);
    expect(await sandbox.exists('-/lit-html@1.1.0')).toBe('dir');
    expect(await sandbox.exists('-/importmap.json')).toBe('file');
    expect(process.exitCode).toBeFalsy();
  });

  test(
    'importmap should add latest as major',
    async () => {
      await sandbox.exec(`add lit-html@1.1.0 --clean`);
      await sandbox.exec(`add lit-html@1.0.0`);
      let importmap: any = await sandbox.read('-/importmap.json');
      importmap = JSON.parse(importmap);
      expect(importmap.imports['lit-html@1']).toBe('/-/lit-html@1.1.0/lit-html.js');
      expect(process.exitCode).toBeFalsy();
    },
    TIMEOUT
  );

  test('prints messages for peers', async () => {
    const out = await sandbox.exec(`add @angular/core@7.0.0 --clean --dry`);
    expect(out).toMatchSnapshot();
    expect(process.exitCode).toBeFalsy();
  });

  test('prints messages when no packages sepecified', async () => {
    const out = await sandbox.exec(`add`);
    expect(out).toContain('No packages specified');
    expect(process.exitCode).toBeFalsy();
  });
});
