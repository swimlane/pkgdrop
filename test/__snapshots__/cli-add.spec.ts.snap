// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add can clean 1`] = `
"Cleaning output directory
Reading existing importmap
Fetching package information for lit-html@1.1.0
Fetching dependecy tree for lit-html@1.1.0
Extracting tarball for lit-html@1.1.0
Writing importmap
Done! [XX s]"
`;

exports[`add can force 1`] = `
"Reading existing importmap
Fetching package information for lit-element@2.1.0
Fetching dependecy tree for lit-element@2.1.0
Extracting tarball for lit-element@2.1.0
Extracting tarball for lit-html@1.4.1
Writing importmap
Done! [XX s]"
`;

exports[`add can force skip on missing packages 1`] = `
"Reading existing importmap
Fetching package information for @swimlane/pkgdrop@0.0.0
Package @swimlane/pkgdrop@0.0.0 not found, skipping
No changes to importmap
Done! [XX s]"
`;

exports[`add displays console messages 1`] = `
"Cleaning output directory
Reading existing importmap
Fetching package information for lit-element@2.1.0
Fetching dependecy tree for lit-element@2.1.0
Extracting tarball for lit-element@2.1.0
Extracting tarball for lit-html@1.4.1
Writing importmap
Done! [XX s]"
`;

exports[`add fail on missing package 1`] = `
"Reading existing importmap
Fetching package information for @swimlane/pkgdrop@0.0.0
Error: Package @swimlane/pkgdrop@0.0.0 not found
	package may be private
	use --force to skip missing packages [XX s]"
`;

exports[`add importmap 1`] = `
"{
  \\"imports\\": {
    \\"lit-element@2.1.0\\": \\"/-/lit-element@2.1.0/lit-element.js\\",
    \\"lit-element@2.1.0/\\": \\"/-/lit-element@2.1.0/\\",
    \\"lit-element@2\\": \\"/-/lit-element@2.1.0/lit-element.js\\",
    \\"lit-element@2/\\": \\"/-/lit-element@2.1.0/\\"
  },
  \\"scopes\\": {
    \\"lit-element@2.1.0/\\": {
      \\"lit-html\\": \\"/-/lit-html@1.4.1/lit-html.js\\",
      \\"lit-html/\\": \\"/-/lit-html@1.4.1/\\"
    },
    \\"lit-html@1.4.1/\\": {}
  }
}"
`;

exports[`add prints messages for peers 1`] = `
"Cleaning output directory [dry run]
Reading existing importmap
Fetching package information for @angular/core@7.0.0
Fetching dependecy tree for @angular/core@7.0.0
@angular/core@7.0.0 requires a peer of rxjs@^6.0.0 but none is installed. You must pkgdrop peer dependencies yourself.
@angular/core@7.0.0 requires a peer of zone.js@~0.8.26 but none is installed. You must pkgdrop peer dependencies yourself.
Extracting tarball for @angular/core@7.0.0 [dry run]
Extracting tarball for tslib@1.14.1 [dry run]
Writing importmap [dry run]
Done! [XX s]"
`;
